src/utils/ServiceLocator.cxxtest

```diff
@@ -0,0 +1,70 @@
+#include "test_main.hxx"
+#include "utils/ServiceLocator.hxx"
+
+using namespace testing;
+
+class SomeService : public RegisterableService<SomeService>
+{
+};
+
+/**
+ * Define a virtual interface that must be implemented, but contains the methods
+ * the consumer wants to use.
+ */
+class OtherInterface : public RegisterableService<OtherInterface>
+{
+public:
+    virtual const char *get_name() = 0;
+};
+
+class OtherImplementation1 : public OtherInterface
+{
+    const char *get_name() override { return "One"; };
+};
+
+class OtherImplementation2 : public OtherInterface
+{
+    const char *get_name() override { return "Two"; };
+};
+
+class ServiceLocatorTests : public Test
+{
+protected:
+    SomeService service_;
+};
+
+TEST_F(ServiceLocatorTests, no_such_service)
+{
+    SomeService *some_service = SomeService::get_service();
+
+    EXPECT_EQ(nullptr, some_service);
+}
+
+TEST_F(ServiceLocatorTests, registered_simple)
+{
+    SomeService::register_service(&service_);
+
+    SomeService *service = SomeService::get_service();
+    EXPECT_EQ(&service_, service);
+
+    OtherInterface *other = OtherInterface::get_service();
+    EXPECT_EQ(nullptr, other);
+}
+
+/**
+ * You can register any implementation of an interface and always retrieve the
+ * registered instance without knowing the class for the registered instance.
+ */
+TEST_F(ServiceLocatorTests, register_via_interface)
+{
+    OtherImplementation1 first;
+    OtherImplementation2 second;
+
+    OtherInterface::register_service(&first);
+    OtherInterface *service = OtherInterface::get_service();
+    EXPECT_EQ("One", service->get_name());
+
+    OtherInterface::register_service(&second);
+    service = OtherInterface::get_service();
+    EXPECT_EQ("Two", service->get_name());
```

Comment:
Please expand this test with the following statements:

1. Show after line 69 how to retrieve the implementation "first".
2. Then destruct second.
3. Then show that "first" can still be retrieved.